### PR TITLE
20 user can apply for job

### PR DESCRIPTION
**Pull request by roscalabrin**
add functionality for a registered user to apply for a job

74 tests:
47 passing and 27 pending

### CODE REVIEW

File: app/services/permission_service.rb [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    return true if controller =="saved_jobs"
+    return true if controller == "users"
+    return true if controller == "sessions"
+    return true if controller == "companies"
+    return true if controller == "sessions"
+    return true if controller == "jobs"
+    return true if controller == "home"
+    return true if controller == "bucket_jobs"
+    return true if controller == "bucket"
+    return true if controller == "job_applications"
```
Comment (roscalabrin) [inline app/services/permission_service.rb:41]: @Laner12 @deborahleehamel this is the line I added to make our redirect work.

File: app/controllers/bucket_jobs_controller.rb [hunk 2]
```diff
@@ -3,16 +3,17 @@ class BucketJobsController < ApplicationController
...
   def create
-    job = Job.find(params[:job_id])
-    @bucket.add_job(job.id)
+    @job = Job.find(params[:job_id])
+    @bucket.add_job(@job.id)
     session[:bucket] = @bucket.contents
-    flash[:success] = "You have started an application for #{job.title}. You have started #{pluralize(@bucket.total_jobs, 'job')} applications."
-    redirect_to bucket_path(job) 
+    flash[:success] = "You have started an application for #{@job.title}. You have started #{pluralize(@bucket.total_jobs, 'job')} applications."
+    create_job_application
```
Comment (roscalabrin) [inline app/controllers/bucket_jobs_controller.rb:10]: @jmejia: here's where we're creating the job application in our database. We haven't implemented yet the functionality to delete the started applications from the database that haven't been completed in a week. Do you see any opportunities to refactor? Is this clear that in the bucketjob controller we create a job application?

File: app/controllers/bucket_jobs_controller.rb [hunk 3]
```diff
@@ -22,16 +23,28 @@ def destroy
   end
 
   private
+    def create_job_application
+    @job_application = @job.job_applications.create(summary: "Please write a brief paragraph explaining why you would be a good fit for this job.")
+    @job_application.update(user_id: current_user.id) if current_user
```
Comment (jmejia) [inline app/controllers/bucket_jobs_controller.rb]: Line 27 creates and 28 updates which is two DB queries. Let's try to get this done with one. I think something like this should work.

``` ruby
@job_application = @job.job_applications.create(
  summary: "Please write a brief paragraph explaining why you would be a good fit for this job.",
  user: current_user
)
```

Comment (jmejia) [inline app/controllers/bucket_jobs_controller.rb]: @roscalabrin ^^

Comment (roscalabrin) [inline app/controllers/bucket_jobs_controller.rb]: good catch! Thanks for the feedback @jmejia!

File: app/controllers/bucket_jobs_controller.rb (hunk 2 above)
Comment (jmejia) [inline app/controllers/bucket_jobs_controller.rb:10]: @roscalabrin There's a good amount of logic that is starting to live here. I'd recommend splitting it out into a PORO. If you're not sure how, go ahead and ship this and create a story to refactor. We can sit down and maybe include anyone that is interested as I live refactor 🙀

File: app/helpers/sessions_helper.rb [hunk 4]
```diff
@@ -3,23 +3,27 @@ def login_referrer
     session[:previous_url].split('/').last if session[:previous_url]
   end
 
+
+  def set_user
+    # if params[:commit]
+      @user = User.find_by(username: params[:session][:username])
+    # else
```
Comment (deborahleehamel) [inline app/helpers/sessions_helper.rb:10]: Let's remove comments in the code before pushing.

**Pull request merged**